qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,431 @@
+                )
+                estimator_result = estimator_job.result()
+                means = np.real(estimator_result.values)
+
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the energy failed!") from exc
+
+            if step > 1:
+                # Compute overlap cost
+                fidelity_job = self.fidelity.run(
+                    [self.ansatz] * len(prev_circs),
+                    prev_circs,
+                    [parameters] * len(prev_circs),
+                    [prev_states[:-1]],
+                )
+                costs = fidelity_job.result().fidelities
+
+                for (state, cost) in zip(range(step - 1), costs):
+                    means += np.real(self.betas[state] * cost)
+
+            if self.callback is not None:
+                metadata = estimator_result.metadata
+                for params, mean, meta in zip([parameters], means, metadata):
+                    self._eval_count += 1
+                    self.callback(self._eval_count, params, mean, meta, step)
+            else:
+                self._eval_count += len(means)
+
+            return means if len(means) > 1 else means[0]
+
+        return evaluate_energy
+
+    @staticmethod
+    def _build_vqd_result() -> VQDResult:
+
+        result = VQDResult()
+        result.optimal_point = []
+        result.optimal_parameters = []
+        result.optimal_value = []
+        result.cost_function_evals = []
+        result.optimizer_time = []
+        result.eigenvalues = []
+        result.optimizer_result = []
+        return result
+
+    @staticmethod
+    def _update_vqd_result(result, opt_result, eval_time, ansatz) -> VQDResult:
+
+        result.optimal_point.append(opt_result.x)
+        result.optimal_parameters.append(dict(zip(ansatz.parameters, opt_result.x)))
+        result.optimal_value.append(opt_result.fun)
+        result.cost_function_evals.append(opt_result.nfev)
+        result.optimizer_time.append(eval_time)
+        result.eigenvalues.append(opt_result.fun + 0j)
+        result.optimizer_result.append(opt_result)
+        return result
+
+
+class VQDResult(VariationalResult, EigensolverResult):
+    """VQD Result."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._cost_function_evals = None
+        self._metadata = {}
+
+    @property
+    def cost_function_evals(self) -> list[int] | None:
+        """Returns number of cost optimizer evaluations"""
+        return self._cost_function_evals
+
+    @cost_function_evals.setter
+    def cost_function_evals(self, value: list[int]) -> None:
+        """Sets number of cost function evaluations"""
+        self._cost_function_evals = value
+
+    @property
+    def metadata(self) -> dict | None:
+        """Returns metadata"""
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, data: dict) -> None:
+        """Sets metadata"""
+        self._metadata = data
```

Comment:
Are we dropping this - I know we talked about it in result but what would this be here given we have multiple values/results. I think we ended up with saying callback would be the way.

Reply:
We are